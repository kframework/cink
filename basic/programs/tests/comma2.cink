int main() {
  int a=1, b=2, c=3, i=0; // comma acts as separator in this line, not as an operator                      ... a=1, b=2, c=3, i=0
  cout << "a=" << a << ", b=" << b << ", c=" << c << ", i=" << i << endl;
	  i = (a, b);             // stores b into i                                                               ... a=1, b=2, c=3, i=2
  cout << "a=" << a << ", b=" << b << ", c=" << c << ", i=" << i << endl;
	  i = a, b;               // stores a into i. Equivalent to (i = a), b;                                    ... a=1, b=2, c=3, i=1
  cout << "a=" << a << ", b=" << b << ", c=" << c << ", i=" << i << endl;
	  i = (a = a + 2, a + b);    // increases a by 2, then stores a+b = 3+2 into i                                ... a=3, b=2, c=3, i=5
  cout << "a=" << a << ", b=" << b << ", c=" << c << ", i=" << i << endl;
	  i = a = a + 2, a + b;      // increases a by 2, then stores a into i. Equivalent to  (i = (a = a + 2)), a + b; ... a=5, b=2, c=3, i=5
  cout << "a=" << a << ", b=" << b << ", c=" << c << ", i=" << i << endl;
	  i = a, b, c;            // stores a into i                                                               ... a=5, b=2, c=3, i=5
  cout << "a=" << a << ", b=" << b << ", c=" << c << ", i=" << i << endl;
	  i = (a, b, c);          // stores c into i                                                               ... a=5, b=2, c=3, i=3
  cout << "a=" << a << ", b=" << b << ", c=" << c << ", i=" << i << endl;
}
