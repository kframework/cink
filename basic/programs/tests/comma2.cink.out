a=1, b=2, c=3, i=0
a=1, b=2, c=3, i=2
a=1, b=2, c=3, i=1
a=3, b=2, c=3, i=5
a=5, b=2, c=3, i=5
a=5, b=2, c=3, i=5
a=5, b=2, c=3, i=3
<T>
    <in>
        ListItem(#buffer ( "" ))
        ListItem(#istream ( 0 ))
    </in>
    <out>
        ListItem(#ostream ( 1 ))
        ListItem(#buffer ( "" ))
    </out>
    <control>
        <k>
            .K
        </k>
        <return>
            noVal
        </return>
        <fstack>
            .List
        </fstack>
        <sideEffects>
            .Bag
        </sideEffects>
        <env>
            main |-> #symScalLoc(1)
        </env>
    </control>
    <genv>
        main |-> #symScalLoc(1)
    </genv>
    <store>
        #symScalLoc(1) |-> lambda (  @ int a ; (a = 1) ; int b ; (b = 2) ;
           int c ; (c = 3) ; int i ; (i = 0) ; (((((((((cout << "a=") << a) <<
           ", b=") << b) << ", c=") << c) << ", i=") << i) << (endl)) ; (i = (
          a , b)) ; (((((((((cout << "a=") << a) << ", b=") << b) << ", c=") 
          << c) << ", i=") << i) << (endl)) ; ((i = a) , b) ; (((((((((cout <<
           "a=") << a) << ", b=") << b) << ", c=") << c) << ", i=") << i) << (
          endl)) ; (i = ((a = (a + 2)) , (a + b))) ; (((((((((cout << "a=") <<
           a) << ", b=") << b) << ", c=") << c) << ", i=") << i) << (endl)) ; 
          ((i = (a = (a + 2))) , (a + b)) ; (((((((((cout << "a=") << a) << 
          ", b=") << b) << ", c=") << c) << ", i=") << i) << (endl)) ; ((i = a
          ) , (b , c)) ; (((((((((cout << "a=") << a) << ", b=") << b) << 
          ", c=") << c) << ", i=") << i) << (endl)) ; (i = (a , (b , c))) ; ((
          (((((((cout << "a=") << a) << ", b=") << b) << ", c=") << c) << 
          ", i=") << i) << (endl)) ; )
        #symScalLoc(2) |-> 5
        #symScalLoc(3) |-> 2
        #symScalLoc(4) |-> 3
        #symScalLoc(5) |-> 3
    </store>
</T>
